docs/PRODUCTION_CONFIGURATION.md
Initially the shift-docs repo is set up for local docker development

## Connection

The production server is hosted on amazon ec2 us-west-2 (oregon) under the shiftgithub@gmail.com amazon account.  @sdboz and @fool know login details for this and are setup to login via ssh via public key.

`ssh ubuntu@api.shift2bikes.org`

It uses an ubuntu image with manually installed docker
The repo is located under

## Git -> production

The git folder is on production is: `/opt/shift-docs/`

In github, whenever we merge a new PR, netlify automatically deploys the code and content to production. The `netlify.toml` controls that under the `[context.production]`. It logs into the production server and uses `./shift pull` to update the server's backend code.

That same shift script can be used on production manually ( ex. `./shift start`, `./shift logs`, etc. as needed. )

## Environment Var Configuration

The `./shift` script looks for a  `./shift.overrides` file to update environment variables ( ex. to set the nginx port to 443 rather than the 4443 port defined at the top of `./shift` ) 

There is no such file in the repo. Instead, there's [shift.overrides.production](https://github.com/shift-org/shift-docs/blob/hosting-docs/shift.overrides.production), and it was manually soft linked as  `/opt/shift-docs/shift.overrides` when setting up the production machine. Nothing in this file is or should be secret. ( ie. no credentials. )

## Images

The event images are served from the `/opt/shift-docs/backend/eventimages` directory, which must have 777 permissions.
Configuration locations:
* `shift.overrides -> shift -> docker-compose.yml -> ./app/config.js`
* `shift.overrides -> shift -> ./app/config.js`
* `./services/nginx/conf.d/shift.conf`

## Mail
See  [secrets.example](https://github.com/shift-org/shift-docs/blob/main/secrets.example). 

A filled out version of that file exists as `/opt/shift-docs/secrets` on our server. I believe `CAL_ADMIN_PASSWORD` is no longer needed, but you'll have to setup the smtp bits so that your sever can send confirmation emails. 

We use amazon's smtp service. ex. `SMTP_HOST=smtp.us-west-2.amazonaws.com` with USER and PASS credentials obtained from the amazon, and `SMTP_DOMAIN=shift2bikes.org`.

## SSL
Our SSL certificate is provided by [lets encrypt](https://letsencrypt.org).  This is required for the hugo front-end running on Netlify to connect to the calendar, and if it doesn't work, you'll see a 500 HTTP error in your devtools (and a blank calendar page in the browser) when trying to fetch calendar data.

The SSL certificate is a bit of a special snowflake in configuration as it lives **OUTSIDE of all the docker containers on the main host** in `/tmp/letsencrypt-auto`, and that is imported into a docker container via [docker-compose.yml](https://github.com/Shift2Bikes/shift-docs/blob/master/docker-compose.yml#L17) and is further [potentially overridden in the `shift` script](https://github.com/Shift2Bikes/shift-docs/blob/master/shift#L14).  Lets Encrypt provides ONLY certificates that last 90 days, so we need to make sure it's updated often. 

Nginx in its docker container **USES** that SSL cert, but **it only reads it on startup, so if the container hasn't been restarted in 90 days, the certificate in use will be expired even if the certificate on disk is up to date**.  

- So, step 0 in debugging SSL failures (for https://api.shift2bikes.org) is to try restarting the docker container:  `cd /opt/shift-docs && ./shift restart nginx`.  
- If that doesn't work, you can try manually running the certificate updater:  `cd /opt/shift-docs && /opt/shift-docs/services/nginx/certbot.sh`.  That should not just get a new certificate but also restart nginx to use it.

That script will show output when run manually, in case something is wrong:

```
Cert not yet due for renewal
Keeping the existing certificate

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Certificate not yet due for renewal; no action taken.
```

or

```
Restarting shift nginx
+ /opt/shift-docs/shift restart nginx
ERROR:
        Can't find a suitable configuration file in this directory or any
        parent. Are you in the right directory?

        Supported filenames: docker-compose.yml, docker-compose.yaml
```

Certbot (which is the name of the package that renews the certificate) logs to a debug log in `/var/log/letsencrypt/letsencrypt.log` that might also be useful to examine for debugging.  

[Vincent](https://github.com/sdboz) and [fool](https://github.com/fool) understand the process thoroughly if you're struggling.


docs/new-server-config-details.txt
AWS config w/Andrew
old instance details:
- t2.micro 
new instance details:
- t3.micro reserved instance reserved for 12 mo on 15 Jan 2020 @ cheaper price (63$/year) - 54.187.123.121 public IP
- additional 2Gb of storage
- keypair shift-server-2020 (fool and Andrew have private key)
instance agnostic details:
- 44.231.45.51 external elastic IP can be mapped to either host

new instance ubuntu setup details:
- stock ubuntu 18.04 LTS 64 bit
- apt-get update ; apt-get upgrade ; apt-get install certbot ; reboot
- josh/andrew/fool ssh keys installed and tested
(not actually sure if docker-compose upgrade is needed anymore? since we are sudo'ing and probably using the /usr/bin version...?  not sure, should test without and see...)
- upgrade docker-compose: curl -L https://github.com/docker/compose/releases/download/1.25.4/docker-compose-`uname -s`-`uname -m` -o /usr/local/bin/docker-compose
- `chmod 777 /var/run/docker.sock` to fix docker-compose permission problems
	(old settings:
	ubuntu@ip-172-31-6-141:/var/run$ ls -ld docker.sock
	srw-rw---- 1 root docker 0 Jan 16 03:57 docker.sock
	)
		- will have to automate this since permissions get reset every reboot
- will have to restore database and eventimages from live host (`cd /opt/shift-docs ; cat backup.mysql | ./shift mysqlpipe` and handle images)

for temporary setup with different hostname:
- set hostname via 3 changes here: https://linuxize.com/post/how-to-change-hostname-on-ubuntu-18-04/
- change hostname to `newshiftapi.shift2bikes.org` in DOMAIN="newshiftapi.shift2bikes.org"
- edit /opt/shift-docs/services/nginx/certbot.sh to include flags to certbot execution: 
	`--agree-tos --email  bikecal@shift2bikes.org`

---- in the end ----
- `shift up` should just work.

---- new work on 12 May 2021 ----
- added /usr/local/bin/certbot.sh from prod, and its associated crontab entry: `28 1 * * * sh /usr/local/bin/certbot.sh 2>&1 > /tmp/certbot-cron-out`.  This worked well!


app/config.js
const path = require('path');
const fs = require("fs");

function env_default(field, def) {
  return process.env[field] ?? def;
}

// node server listen for its http requests
// fix? there is no such environment variable right now.
const listen = env_default('NODE_PORT', 3080);

// the user facing server.
const siteHost = siteUrl(listen);

// location of app.js ( same as config.cs )
const appPath =  path.resolve(__dirname);

// for max file size
const bytesPerMeg = 1024*1024;

const staticFiles = env_default('SHIFT_STATIC_FILES');

const config = {
  appPath,
  api: {
    header: 'Api-Version',
    version: "3.56.0",
  },
  db: {
    host: env_default('MYSQL_HOST', 'db'),
    port: 3306, // standard mysql port.
    user: env_default('MYSQL_USER', 'shift'),
    pass: env_default('MYSQL_PASSWORD', 'ok124'),
    name: env_default('MYSQL_DATABASE', 'shift'),
    type: "mysql2", // name of driver, installed by npm
  },
  // a nodemailer friendly config, or false if smtp is not configured.
  smtp: getSmtpSettings(),
  site: {
    name: "SHIFT to Bikes",
    listen,
    path: "/",
    // used for crawl url, shareable links, and the manage url sent in email
    url(...parts) {
     const base = `${siteHost}${config.site.path}`;
     return base + parts.join("/");
    },
    helpPage() {
      return config.site.url("pages/calendar-faq/");
    },
    staticFiles,
    devEndpoints: !staticFiles ? null : [{
      // ex. http://localhost:3080/addevent/edit-1-d00c888b0a1d4bab8107ba2fbe2beddf
      // loads http://localhost:3080/addevent/index.html
      url: "/addevent/edit-:series-:secret",
      filePath: path.posix.resolve(staticFiles, 'addevent', 'index.html')
    }, {
      // ex. http://localhost:3080/calendar/event-201
      // loads http://localhost:3080/calendar/event/index.html
      url: "/calendar/event-:caldaily",
      filePath: path.posix.resolve(staticFiles, 'calendar/event', 'index.html')
    }, {
      // ex. http://localhost:3080/events/201
      // loads http://localhost:3080/events/index.html
      url: "/events/:caldaily_id",
      filePath: path.posix.resolve(staticFiles, 'events', 'index.html')
    }],
  },
  // various useful email addresses
  // ( for sendConfirmationEmail() )
  // https://nodemailer.com/message/addresses/
  email: {
    sender: {
      name: 'SHIFT to Bikes',
      address: 'bikefun@shift2bikes.org'
    },
    // the confirmation emailer sets this as the reply-to
    support: "bikecal@shift2bikes.org",
    // the confirmation emailer blind copies this address
    moderator: "shift-event-email-archives@googlegroups.com",
    logfile: function() {
      const logfile = env_default('SHIFT_EMAIL_LOG');
      return logfile ? path.resolve(appPath, logfile) : false;
    }()
  },
  image: {
    // storage location for event images
    dir: path.resolve(appPath,
        env_default('SHIFT_IMAGE_DIR', "eventimages")),
    // used for generated event image links
    // ( see also the ngnix config )
    path: env_default('SHIFT_IMAGE_PATH', "/eventimages/"),
    // used for event image links.
    // for backwards compatibility:
    // 1. if you pass a falsey value, this returns null.
    // 2. php used host relative image links so this does too.
    // ex. not "https://shift2bikes.org/eventimages/9248-2.png"
    // but "/eventimages/9248-2.png"
    url(name) {
      return !name ? null : path.posix.join(config.image.path, name);
    },
    // see also: shift.conf (ngnix configuration)
    // we set this *larger* than ngnix so to accept everything it gives us.
    // ( while still having a fail-safe if something unexpected occurs. )
    maxFileSize: 5.5 * bytesPerMeg,
    // mime type to a desired image extension.
    // we assign the extension based on the content.
    imageTypes:{
      'image/gif'  : '.gif',
      'image/jpeg' : '.jpg',
      'image/pjpeg': '.jpg',
      'image/png'  : '.png',
    },
  },
  crawl: {
    image: 'https://www.shift2bikes.org/images/shiftLogo_plain.gif',
    title: 'Shift/Pedalpalooza Calendar',
    description: `Find fun bike events and make new friends!` +
      `Shift helps groups and individuals to promote their "bike fun" events.`,
  },
  cal: {
    pedalp: {
      name: 'Pedalpalooza Bike Calendar',
      desc: 'Find fun Pedalpalooza bike events!',
      guid: 'shift@shift2bikes.org',
      filename: 'pedalpalooza-calendar',
    },
    shift: {
      name: 'Shift Community Calendar',
      desc: 'Find fun bike events all year round.',
      guid: 'community@shift2bikes.org',
      filename: 'shift-calendar',
    },
    // shared properties:
    base: {
      ext: '.ics',
      maxage: 60*60*3, // 3 hours
      // the software that created the calendar
      prod: '-//shift2bikes.org//NONSGML shiftcal v2.1//EN',
    },
  },
};
module.exports = config;

// determine the user facing prefix for generating page and image links.
// production         - https://shift2bikes.org
// docker dev         - https://localhost:4443
// standalone or test - http://localhost:3080
function siteUrl(proxyPort) {
  const hostname = env_default('SHIFT_DOMAIN', 'localhost');
  // production uses 443, docker dev uses 4443, standalone is null.
  const serverPort = env_default('NGINX_HTTPS_PORT', null);
  const protocol = serverPort ? "https://" : "http://";
  const portstr  = (serverPort === '443') ? '' : ':' + (serverPort ?? proxyPort);
  return protocol + hostname + portstr;
}

// returns a nodemailer friendly config, or false if smtp is not configured.
function getSmtpSettings() {
  // hack to read email configuration from a file;
  // tools/newEtherealConfig.js can generated it.
  const emailCfg = env_default('SHIFT_EMAIL_CFG');
  if (emailCfg) {
    try {
      const raw= fs.readFileSync(emailCfg, "utf8");
      return JSON.parse(raw);
    } catch (err) {
      // its okay if there is no such file...
      if (err.code !== 'ENOENT') {
        throw(err);
      }
    }
  }
  // assumes that if SMTP_HOST is set, the rest is okay.
  const host = env_default('SMTP_HOST');
  if (!host) {
    return false;
  } else {
    // assumes that if SMTP_HOST is set, the rest is okay;
    // ( and returns a nodemailer config: https://nodemailer.com/smtp/ )
    return {
      host: host,
      port:  587,
      // secure should be true for 465;
      // false for everything else; and everyone seems to want 587.
      secure: false,
      auth: {
        user: env_default('SMTP_USER'),
        pass: env_default('SMTP_PASS'),
      }
    };
  }
}
